utility/map_str_enhancer.go

```diff
@@ -26,101 +26,88 @@ import (
 	"github.com/elastic/beats/libbeat/common"
 )
 
-func Get(m common.MapStr, key string) interface{} {
-	ret, _ := m.GetValue(key)
-	return ret
+// Set takes a map and changes key to point to the provided value.
+// In case the provided value is nil or of length 0, the key is deleted from the map .
+func Set(m common.MapStr, key string, val interface{}) {
+	update(m, key, val, true)
 }
 
-func DeepAdd(m common.MapStr, dottedKeys string, val interface{}) {
-	keys := strings.Split(dottedKeys, ".")
-	if len(keys) == 0 {
-		return
-	}
-	reverse(keys)
-	v := val
-	for _, k := range keys {
-		subMap := common.MapStr{}
-		Add(subMap, k, v)
-		v = subMap
-	}
-	m.DeepUpdate(v.(common.MapStr))
+// Update takes a map and changes key to point to the provided value.
+// In case the provided value is nil nothing is changed.
+func Update(m common.MapStr, key string, val interface{}) {
+	update(m, key, val, false)
 }
 
-func reverse(slice []string) {
-	size := len(slice)
-	for i := 0; i < len(slice)/2; i++ {
-		slice[i], slice[size-i-1] = slice[size-i-1], slice[i]
-	}
-}
-
-func Add(m common.MapStr, key string, val interface{}) {
+func update(m common.MapStr, key string, val interface{}, remove bool) {
 	if m == nil || key == "" {
 		return
 	}
 
 	if val == nil {
-		delete(m, key)
+		if remove {
+			delete(m, key)
+		}
 		return
 	}
 
 	switch value := val.(type) {
 	case *bool:
 		if value != nil {
 			m[key] = *value
-		} else {
+		} else if remove {
```

Comment:
are these elses reachable given the check on line 46?

Reply:
Line 46 is just a shortcut for cases where the `val` itself is `nil`. The `else` checks are for the cases where the interface `val` is not nil, meaning that it holds a type information and a pointer to the data. The data themselves can be nil again though. 

I found https://blog.golang.org/laws-of-reflection#TOC_3 very helpful when I ran into this.

Reply:
Ah, that makes sense, thanks